File: algorithms/atoi.cpp
#include <iostream>
#include <cstdint>
#include <cstring>
#include <stdexcept>

int my_atoi( char* buffer, unsigned int length )
{
  if (length == 0)
  {
    throw std::runtime_error("invalid length " + std::to_string(length));
  }

  int result = 0;
  unsigned int i = 0;

  bool is_negative = buffer[0] == '-';

  if (is_negative)
  {
    if (length == 1)
    {
      throw std::runtime_error("invalid string " + std::string(buffer));
    }

    i++;
  }

  while(i < length)
  {
    result *= 10;

    if (buffer[i] < '0' || buffer[i] > '9')
    {
      throw std::runtime_error("invalid character in column " + std::to_string(i));
    }

    result += buffer[i] - '0';

    if (result < 0)
    {
      throw std::runtime_error("overflow detected");
    }

    i++;
  }

  return is_negative ? -1 * result : result;
}

int main(int argc, char** argv)
{
  if (argc != 2)
  {
    std::cout << "usage: " << argv[0] << " <integer>" << std::endl;
    return -1;
  }

  try
  {
    std::cout << my_atoi(argv[1], std::strlen(argv[1])) << std::endl;
  }
  catch( std::runtime_error & e)
  {
    std::cout << e.what() << std::endl;
  }

  return 0;
}

File: python/exercises/other/atoi.py
import sys

def atoi(integer_string):
    result = 0
    i = 0
    is_negative = integer_string[0] == "-"

    if is_negative:
        if len(integer_string) == 1:
            raise Exception("Invalid integer string")
        i += 1

    while i < len(integer_string):
        result *= 10

        char = ord(integer_string[i])

        if char < ord("0") or char > ord("9"):
            raise Exception("Invalid character {} in column {}".format(char, i))

        result += char - ord("0")

        i += 1

    if is_negative:
        result *= -1

    return result

if __name__ ==  "__main__":
    if len(sys.argv) != 2:
        print("usage: python -m atoi <integer>")
        sys.exit(0)

    result = atoi(sys.argv[1])

    print("integer value: {}".format(result))
